### PR TITLE
PARQUET-595: API for KeyValue metadata

### PR DESCRIPTION
**Pull request by wesm**
This supersedes #309 and incorporates the `std::shared_ptr<const KeyValueMetadata>` pattern so less copying is needed in Parquet for metadata inbound from Arrow (and vice versa).

close #309

### CODE REVIEW

File: src/parquet/arrow/schema.h [hunk 1]
```diff
@@ -44,13 +45,17 @@ ::arrow::Status PARQUET_EXPORT NodeToField(
...
-    const std::vector<int>& column_indices, std::shared_ptr<::arrow::Schema>* out);
+    const std::vector<int>& column_indices,
+    const std::shared_ptr<const KeyValueMetadata>& key_value_metadata,
+    std::shared_ptr<::arrow::Schema>* out);
 
-::arrow::Status PARQUET_EXPORT FromParquetSchema(
-    const SchemaDescriptor* parquet_schema, std::shared_ptr<::arrow::Schema>* out);
+::arrow::Status PARQUET_EXPORT FromParquetSchema(const SchemaDescriptor* parquet_schema,
+    const std::shared_ptr<const KeyValueMetadata>& key_value_metadata,
+    std::shared_ptr<::arrow::Schema>* out);
```
Comment (advancedxy) [inline src/parquet/arrow/schema.h]: Interface question: 
I think it's better to keep the previous exported interface.
```
::arrow::Status PARQUET_EXPORT FromParquetSchema(
      const SchemaDescriptor* parquet_schema, std::shared_ptr<::arrow::Schema>* out);
```

so it doesn't break user's code and if user doesn't care about key_value_metadata, he can just ignore them.

what do you think? @wesm

Comment (wesm) [inline src/parquet/arrow/schema.h]: I agree. I will make a pass through and add backwards compatible APIs.

Comment (wesm) [inline src/parquet/arrow/schema.h]: Done. Does this patch look OK?

File: src/parquet/arrow/schema.h [hunk 2]
```diff
@@ -19,6 +19,7 @@
 #define PARQUET_ARROW_SCHEMA_H
 
 #include <memory>
+#include <string>
```
Comment (cpcloud) [inline src/parquet/arrow/schema.h]: `std::string` isn't used in this file right?

Comment (wesm) [inline src/parquet/arrow/schema.h]: Removed. We could use a include-what-you-use pass on the whole codebase

File: src/parquet/arrow/writer.cc [hunk 3]
```diff
@@ -18,6 +18,7 @@
 #include "parquet/arrow/writer.h"
 
 #include <algorithm>
+#include <string>
```
Comment (cpcloud) [inline src/parquet/arrow/writer.cc]: Is `std::string` being used in this file?

File: src/parquet/file/writer.h [hunk 4]
```diff
@@ -103,12 +124,14 @@ class PARQUET_EXPORT ParquetFileWriter {
...
-      const std::shared_ptr<WriterProperties>& properties = default_writer_properties());
+      const std::shared_ptr<WriterProperties>& properties = default_writer_properties(),
+      const std::shared_ptr<const KeyValueMetadata>& key_value_metadata = nullptr);
 
   static std::unique_ptr<ParquetFileWriter> Open(
       const std::shared_ptr<OutputStream>& sink,
       const std::shared_ptr<schema::GroupNode>& schema,
-      const std::shared_ptr<WriterProperties>& properties = default_writer_properties());
+      const std::shared_ptr<WriterProperties>& properties = default_writer_properties(),
+      const std::shared_ptr<const KeyValueMetadata>& key_value_metadata = nullptr);
```
Comment (cpcloud) [inline src/parquet/file/writer.h]: If someone is using this API, and they're linking against the `libparquet` shared object, will changing this signature still allow existing code to continue working?

File: src/parquet/file/writer.h [hunk 5]
```diff
@@ -20,8 +20,10 @@
 
 #include <cstdint>
 #include <memory>
+#include <string>
```
Comment (cpcloud) [inline src/parquet/file/writer.h]: Is `std::string` being used here?

File: src/parquet/file/writer-internal.h [hunk 6]
```diff
@@ -19,6 +19,7 @@
 #define PARQUET_FILE_WRITER_INTERNAL_H
 
 #include <memory>
+#include <string>
```
Comment (cpcloud) [inline src/parquet/file/writer-internal.h]: `string` used in this file?

File: src/parquet/file/writer.h (hunk 4 above)
Comment (wesm) [inline src/parquet/file/writer.h]: Yep, I think this actually will generate two symbols, one with the new argument and one without

File: src/parquet/file/writer-internal.h (hunk 6 above)
Comment (wesm) [inline src/parquet/file/writer-internal.h]: removed and above, artifact of the old data structure

**Comment by cpcloud** [comment]
Thanks for getting the `shared_ptr` stuff up! I'm cranking on ARROW-881 as we speak.

**Comment by cpcloud** [comment]
(Along with decimal madness)

**Comment by advancedxy** [comment]
I did another quick review. LGTM.

**Comment by wesm** [comment]
Thanks. Merging

**Pull request closed without merging**